scripts/lib/mkcloud-libvirt.sh

```diff
@@ -57,108 +57,285 @@ function libvirt_net_start()
         ip link set mtu 9000 dev $dev
     done
 
-    boot_mkcloud=/etc/init.d/boot.mkcloud
-    boot_mkcloud_d="$boot_mkcloud.d"
-    boot_mkcloud_d_cloud="$boot_mkcloud_d/$cloud"
-
-    if [ -z "$NOSETUPPORTFORWARDING" ] ; then
-        # FIXME: hardcoded assumptions about admin net host range
-        nodehostips=$(seq -s ' ' 81 $((80 + $nodenumber)))
-
-        : ${cloud_port_offset:=1100}
-        mosh_start=$(( $cloud_port_offset + 60001 ))
-        mosh_end=$((   $cloud_port_offset + 60010 ))
-
-        mkdir -p $boot_mkcloud_d
-        cat > $boot_mkcloud_d_cloud <<EOS
-#!/bin/bash
-# Auto-generated from $0 on `date`
-
-iptables_unique_rule () {
-    # First argument must be chain
-    if iptables -C "\$@" 2>/dev/null; then
-        echo "iptables rule already exists: \$*"
-    else
-        iptables -I "\$@"
-        echo "iptables -I \$*"
-    fi
+    onhost_setup_portforwarding
 }
 
-# Forward ports to admin server
-for port in 22 80 443 3000 4000 4040; do
-    iptables_unique_rule PREROUTING -t nat -p tcp \\
-        --dport \$(( $cloud_port_offset + \$port )) \\
-        -j DNAT --to-destination $adminip:\$port
-done
-
-# Connect to admin server with mosh (if installed) via:
-#   mosh -p $mosh_start --ssh="ssh -p $(( $cloud_port_offset + 22 ))" `hostname -f`
-iptables_unique_rule PREROUTING -t nat -p udp \\
-    --dport $mosh_start:$mosh_end \\
-    -j DNAT --to-destination $adminip
-
-# Forward ports to non-admin nodes
-for port in 22 80 443 5000 7630; do
-    for host in $nodehostips; do
-        # FIXME: hardcoded assumptions about admin net host range
-        offset=80
-        host_port_offset=\$(( \$host - \$offset ))
-        iptables_unique_rule PREROUTING -t nat -p tcp \\
-            --dport \$(( $cloud_port_offset + \$port + \$host_port_offset )) \\
-            -j DNAT --to-destination $net_admin.\$host:\$port
-    done
-done
+function libvirt_prepare()
+{
+    # libvirt
+    libvirt_modprobe_kvm
+    libvirt_start_daemon
 
-iptables_unique_rule PREROUTING -t nat -p tcp --dport 6080 \\
-    -j DNAT --to-destination $net_public.2
+    # network
+    ${mkcloud_lib_dir}/libvirt/net-config $cloud $cloudbr $admingw $adminnetmask $cloudfqdn $adminip $forwardmode > /tmp/$cloud-admin.net.xml
+    ${mkcloud_lib_dir}/libvirt/net-start /tmp/$cloud-admin.net.xml || exit $?
+    libvirt_net_start
+}
+
+function libvirt_do_setupadmin()
```

Comment:
so basically an "interface" in scripts/mkcloud that starts with "do_foo" and calls out ${mkclouddriver}_do_foo ? works for me.